File: leetcode/138-deepcopy-linkedlist.java


class Node {
    int val;
    Node next;
    Node random;

    Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return null;
        Map<Node, Node> map = new HashMap<>();
        Node curr = head;
        // first pass -> create and primitive
        while (curr != null) {
            map.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        // second pass - link to newly created 
        curr = head;
        while (curr != null) {
            map.get(curr).next = map.get(curr.next);
            map.get(curr).random = map.get(curr.random);
        }

        // return 
        return map.get(head);
    }
}

File: leetcode/138-deepcopy-linkedlist.js

const copyRandomList = (head) => {
    if (!head) return null;
    const mapper = new Map();
    let curr = head;
    while (curr) {
        mapper.set(curr, new Node(curr.val));
        curr = curr.next;
    }

    curr = head;
    while (curr) {
        mapper.get(curr).next = mapper.get(curr.next) || null;
        mapper.get(curr).random = mapper.get(curr.random) || null;
        curr = curr.next;
    }
    return mapper.get(head);
}